docs/architecture/ADR-002-deposit-assets.md

```diff
@@ -0,0 +1,117 @@
+
+The Asset Facility will have the capability to mint and burn uTokens (but not their corresponding original asset types). It will have access to an allow-list of said asset and uToken types.
+
+The Asset Facility will possess a `Module Account` to store original assets, and that module account should be forbidden from being the recipient of transactions except those intended by the module (see the warning on [this page](https://docs.cosmos.network/master/modules/bank/).)
+
+The Asset Facility should harness the Cosmos `banking` module's `BaseKeeper` for the following capabilities
+- Read the supply (chain-wide) of a given coin (both base assets and uTokens)
+- Send coins (both base assets and uTokens) from module account to user account (and vice versa)
+- Mint uTokens
+- Burn uTokens
+- `SendKeeper`: Use BlockedAddr feature to guard against unexpected transfers to module account(s)
+- `ViewKeeper`: Read individual account balances
+
+### Basic Message Types
+
+For reference, here is the `Bank` module's built in coin transfer message as seen [here](https://docs.cosmos.network/v0.39/basics/app-anatomy.html), which is used when regular users send tokens to one another:
+```go
+// MsgSend - high level transaction of the coin module
+type MsgSend struct {
+	FromAddress sdk.AccAddress `json:"from_address" yaml:"from_address"`
+	ToAddress   sdk.AccAddress `json:"to_address" yaml:"to_address"`
+	Amount      sdk.Coins      `json:"amount" yaml:"amount"`
+}
+```
+The `sdk.Coins` type is a slice (ordered list) of `sdk.Coin` which contain a token type and amount.
+
+To implement the deposit functionality of the Asset Facility, the two common message types will be:
+```go
+// MsgDeposit - a user wishes to deposit assets and receive uAssets
+type MsgDeposit struct {
+	FromAddress sdk.AccAddress `json:"from_address" yaml:"from_address"`
+	Amount      sdk.Coins      `json:"amount" yaml:"amount"`
+}
+// MsgWithdraw - a user wishes to redeem uAssets for original assets
+type MsgWithdraw struct {
+	ToAddress sdk.AccAddress `json:"to_address" yaml:"to_address"`
+	Amount      sdk.Coins      `json:"amount" yaml:"amount"`
+}
+```
+This resembles the built-in MsgSend, but either ToAddress or FromAddress is removed because the module's address should be used automatically.
+
+MsgDeposit must use only whitelisted, non-uToken denominations. MsgWithdraw must use only uToken denominations.
+
+These messages should trigger the appropriate reaction (disbursement of uTokens after deposit, return of assets on withdrawal). The exchange rate defined in ADR-001 must be used.
+
+_Note: The `Coins` type seen in the `Amount` fields can contain multiple token types. Deposits and withdrawals should fail if even one of the coin types fails or is invalid, rather than partially succeeding._
+
+It is necessary that `MsgDeposit` and `MsgWithdrawal` be signed by the owner's account. According to the [Transactions Page](https://docs.cosmos.network/master/core/transactions.html)
```

Comment:
```suggestion
MsgDepositAsset must use only whitelisted, non-uToken denominations. MsgWithdrawAsset must use only uToken denominations.

These messages should trigger the appropriate reaction (disbursement of uTokens after deposit, return of assets on withdrawal). The exchange rate defined in ADR-001 must be used.

_Note: The `Coins` type seen in the `Amount` fields can contain multiple token types. Deposits and withdrawals should fail if even one of the coin types fails or is invalid, rather than partially succeeding._

It is necessary that `MsgDepositAsset` and `MsgWithdrawAsset` be signed by the owner's account. According to the [Transactions Page](https://docs.cosmos.network/master/core/transactions.html)
```